### ISSUE TITLE
Add at-nest selector support

### ISSUE THREAD

**Issue by woble** [stylelint/stylelint#1339]
> Describe the issue. Is it a bug or a feature request (new rule, new option, etc.)?

When linting nested selectors using [postcss-nesting](https://github.com/jonathantneal/postcss-nesting) syntax, stylelint does not produce any warnings.

> Which rule, if any, is this issue related to?

`selector-max-compound-selectors`

> What CSS is needed to reproduce this issue?

```css
.foo {
    color: red;

    @nest .baz {
        color: tomato;
    }
}
```

> What stylelint configuration is needed to reproduce this issue?

```json
{
  "rules": {
    "selector-max-compound-selectors": 1
  }
}
```

> Which version of stylelint are you using?

`6.5.0`

> How are you running stylelint: CLI, PostCSS plugin, Node API?

- CLI as `stylelint test.pcss`
- gulp-stylelint
- SublimeLinter-contrib-stylelint

> Does your issue relate to non-standard syntax (e.g. SCSS, nesting, etc.)?

Yes. Proposed [CSS Nesting Module Level 3](http://tabatkins.github.io/specs/css-nesting/) specification.

> What did you expect to happen?

A warning at `@nest .baz {`

> What actually happened (e.g. what warnings or errors you are getting)?

No warning or errors.

**Comment by jeddy3** [comment]
I'm not sure this is valid CSS:

```css
```css
.foo {
    color: red;

    @nest .baz {
        color: tomato;
    }
}
```

From the spec: "The only difference is that the selector used in a `@nest` rule must be **nest-containing**, which means it contains a nesting selector in it somewhere".

Can you try (with the *nesting selector*):

```css
.foo {
    color: red;

    @nest & .baz {
        color: tomato;
    }
}
```

Or, try using *direct nesting*:

```css
.foo {
    color: red;

    & .baz {
        color: tomato;
    }
}
```

And then report back.

**Comment by woble** [comment]
My mistake, it should be the following indeed:

```css
.foo {
    color: red;

    @nest & .baz {
        color: tomato;
    }
}
```

But that doesn't work either.

We use `@nest` for consitency sake, e.g. `@nest & .baz {}` and `@nest .baz & {}` vs `& .baz {}` and `@nest .baz & {}`.

In fact, `.foo { @nest .baz & {} }` should produce a warning as well when `selector-max-compound-selectors` is set to `1` for example.

**Comment by evilebottnawi** [comment]
@woble close?

**Comment by jeddy3** [comment]
Thanks for trying with the *nesting selector*.

> In fact, .foo { @nest .baz & {} } should produce a warning as well when selector-max-compound-selectors is set to 1 for example.

Yes it should. 

> @woble close?

Nope, it doesn't appear to be working even with the correct syntax. We need to investigate further.

**Comment by jeddy3** [comment]
I got a bit side-tracked with this issue... anyway...

Most, if not all, of the selector rules don't support the `@nest` syntax, as we generally walk through just rules and check their selectors. I don't think we check [`@nest` selectors](https://tabatkins.github.io/specs/css-nesting/#at-nest) in any rule.

I've a [rough PoC](https://github.com/stylelint/stylelint/pull/1343) to show how we might go about adding support to this rule and `selector-max-specificity` (as they are very similar). What we need to discuss over there is if and how we're going to address this across all the selector rules.

@woble Thanks, as always, for the in-depth report. However we go about this, it'll have to wait until `7.0` as it will be a breaking change. Sorry about that. As an aside, this is interesting to know:

> We use `@nest` for consitency sake, e.g. `@nest & .baz {}` and `@nest .baz & {}` vs `& .baz {}` and `@nest .baz & {}`.

With my current team we do the opposite i.e. we use `@nest` *to signal* that something more complicated than direct nesting is about to happen e.g.

```css
.foo {
  prop: value;

  /* simple pseudo or down-the-tree nesting (less prone to side-effects) */
  &:hover {}
  & .baz {}

  /* ey-up, we're going up-the-tree... we should probably be careful when doing this.*/
  @nest .stateClassOnParentSomewhere & {}
}
```

**Comment by jeddy3** [comment]
> My mistake, it should be the following indeed:

@woble As an another aside, there's a [proposal](https://github.com/stylelint/stylelint/issues/962) for a new rule that would help you avoid those erroneously missing *nesting selectors*. The proposal is looking for someone to champion it, i.e. build and maintain it, if you're interested?

**Comment by jeddy3** [comment]
Closing. Ref: https://github.com/stylelint/stylelint/pull/1343#issuecomment-225473367
